lib/sync_engine/abstract/user_manager.rb

```diff
@@ -38,7 +38,14 @@ def change_pw(user, password, params, user_agent = '')
       user.encrypted_password = hash_password(password)
       user.update!(registration_params(params))
 
-      create_session(user, params, user_agent, upgrading_protocol_version)
+      # We want to create a new session only if upgrading from a protocol version that does not
+      # support sessions (i.e: 003 to 004) or if not upgrading protocols at all (i.e: change password).
+      if (upgrading_protocol_version && user.version.to_i == 4) || !upgrading_protocol_version
```

Comment:
The number `4` shouldn't be hardcoded. Should be defined as a constant, like 'LastNonSessionedProtocolVersion` or something like that.

Reply:
Or could be a User method. `user.uses_sessions` or something like that.

Reply:
> or if not upgrading protocols at all (i.e: change password).

Why would we create a new session if just changing the password? Even if that were the correct logic (which I don't believe it is), what happens to the old session the user had?

Reply:
> The number `4` shouldn't be hardcoded. Should be defined as a constant, like 'LastNonSessionedProtocolVersion` or something like that.

There's the [supports_sessions?](https://github.com/standardnotes/syncing-server/blob/97da74815ac62e8658ac8e9121187cff1e07b4b5/app/models/user.rb#L124) method. Currently, it is assuming we will support sessions beyond version `004`, so I guess we shouldn't make any assumptions and correct it to `version.to_i == 4`.

I would also argument that an upgrade from `004` to any future version will most likely not require to create a new session. This is only required if switching from `003` to `004`.

Reply:
> Why would we create a new session if just changing the password? Even if that were the correct logic (which I don't believe it is), what happens to the old session the user had?

If a user with version `003` is just changing their password, they will need a new JWT. In this case, `create_session` will return [here](https://github.com/standardnotes/syncing-server/blob/97da74815ac62e8658ac8e9121187cff1e07b4b5/lib/sync_engine/abstract/user_manager.rb#L117).

Perhaps we should rename `create_session` to something that describes both JWT and Sessions.

Reply:
> If a user with version 003 is just changing their password

Changing a password always upgrades their protocol version. User doesn't need to explicitly choose to upgrade. If a user on 003 changes their password, they are upgraded.

> if (upgrading_protocol_version && user.version.to_i == 4)

This should be `if (upgrading_protocol_version && new_protocol_version == 4) ` to be more explicit.

Reply:
You're still not explaining this: `if not upgrading_protocol_version then create new session`. Why would we create a new session on a normal password change?

Reply:
> Changing a password always upgrades their protocol version. User doesn't need to explicitly choose to upgrade. If a user on 003 changes their password, they are upgraded.

Got it :+1: 

> You're still not explaining this: `if not upgrading_protocol_version then create new session`. Why would we create a new session on a normal password change?

Given that the user is on the latest protocol version, and they wish to change their password, then there's no need to create a new session.

Reply:
> Given that a user is on the latest protocol version, and they wish to change their password, then there's no need to create a new session.

Correct. However this is not what the code is doing.

Reply:
`|| !upgrading_protocol_version` should be removed.